src/compiler/transformers/esDecorators.ts

```diff
@@ -0,0 +1,2087 @@
+
+            if (classInfo.instanceExtraInitializersName && !getFirstConstructorWithBody(node)) {
+                const initializerStatements = prepareConstructor(node, classInfo);
+                if (initializerStatements) {
+                    const extendsClauseElement = getEffectiveBaseTypeNode(node);
+                    const isDerivedClass = !!(extendsClauseElement && skipOuterExpressions(extendsClauseElement.expression).kind !== SyntaxKind.NullKeyword);
+                    const constructorStatements: Statement[] = [];
+                    if (isDerivedClass) {
+                        const spreadArguments = factory.createSpreadElement(factory.createIdentifier("arguments"));
+                        const superCall = factory.createCallExpression(factory.createSuper(), /*typeArguments*/ undefined, [spreadArguments]);
+                        constructorStatements.push(factory.createExpressionStatement(superCall));
+                    }
+
+                    addRange(constructorStatements, initializerStatements);
+
+                    const constructorBody = factory.createBlock(constructorStatements, /*multiLine*/ true);
+                    syntheticConstructor = factory.createConstructorDeclaration(/*modifiers*/ undefined, [], constructorBody);
+                }
+            }
+
+            // Used in steps 5, 7, and 11
+            if (classInfo.staticExtraInitializersName) {
+                classDefinitionStatements.push(
+                    createLet(classInfo.staticExtraInitializersName, factory.createArrayLiteralExpression())
+                );
+            }
+
+            // Used in steps 6, 8, and during construction
+            if (classInfo.instanceExtraInitializersName) {
+                classDefinitionStatements.push(
+                    createLet(classInfo.instanceExtraInitializersName, factory.createArrayLiteralExpression())
+                );
+            }
+
+            // Used in steps 7, 8, 12, and construction
+            if (classInfo.memberInfos) {
+                forEachEntry(classInfo.memberInfos, (memberInfo, member) => {
+                    if (isStatic(member)) {
+                        classDefinitionStatements.push(createLet(memberInfo.memberDecoratorsName));
+                        if (memberInfo.memberInitializersName) {
+                            classDefinitionStatements.push(createLet(memberInfo.memberInitializersName, factory.createArrayLiteralExpression()));
+                        }
+                        if (memberInfo.memberDescriptorName) {
+                            classDefinitionStatements.push(createLet(memberInfo.memberDescriptorName));
+                        }
+                    }
+                });
+            }
+
+            // Used in steps 7, 8, 12, and construction
+            if (classInfo.memberInfos) {
+                forEachEntry(classInfo.memberInfos, (memberInfo, member) => {
+                    if (!isStatic(member)) {

```

Comment:
why split the static/non-static additions into two separate loops? I guess it makes the generated code cleaner to have static/non-static separate?

Reply:
That's pretty much it, yes. My choices are, essentially:
1. Document order (matches what's written)
2. Statics then non-statics (keeps all of the `_static_`-prefixed variables grouped)
3. Static methods, then instance methods, then static fields, then instance fields (matches spec application order)

I went with (2) as being the most readable.